### DISCUSSION TITLE
eth_getBlockByHash doesn't provide correct block hash.

### DISCUSSION THREAD

**Discussion by scalalang2** [0xPolygon/polygon-edge#679] [General]
### Description

Hi,
I've been stuck in trouble for integration with polygon edge.
I made a jsonrpc call `eth_getBlockByHash` with [go-ethereum](https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go#L44) client code. 
Then it gives me an incorrect block hash.

### Your environment
* polygon edge mainnet

### Steps to reproduce

Please execute following test code.
This calls official polygon edge network and try to retreive block information.
Then it gives me an incorect `block hash`.

```golang
func TestBlockByHash(t *testing.T) {
	edgeNet := "https://rpc-edgenet.polygon.technology"
	// https://explorer-edgenet.polygon.technology/block/1320009/transactions
	blockHash := "0xd4f333b237f56436f6e3074b22a1e0dada6dff4ee574778fa5dcdcebe118323d"
	client, err := ethclient.DialContext(context.Background(), edgeNet)
	if err != nil {
		t.Errorf("failed to connect to the network")
	}

	block, err := client.BlockByHash(context.Background(), ethcommon.HexToHash(blockHash))
	if err != nil {
		t.Errorf("failed to fetch block")
	}

	if block.Hash().Hex() != blockHash {
		t.Errorf("received block is not eqaul, query: %s, got: %s", blockHash, block.Hash().Hex())
	}
}
```

The result of this code looks as below
```
received block is not eqaul, query: 0xd4f333b237f56436f6e3074b22a1e0dada6dff4ee574778fa5dcdcebe118323d, got: 0xd3346a9a759505837601e94730c977b4d7697d13469c248993cb6a96b6b5845e
```

It does not seem to be a json rpc protocol problem, 
It seems that the block schema of polygon edge do not fit into ethereum one.

**Comment by 0xAleksaOpacic** [comment]
hi @scalalang2 
Thank you for opening this issue, I need to investigate to find the cause of this issue. 
Once I have more detail than I will notify you :)

**Comment by 0xAleksaOpacic** [comment]
Hi @scalalang2 
I finally know what is the root cause of this issue. The problem is that Polygon Edge is calculating block hash differently than Ethereum does.
To be precise:
1. We are missing 2 fields when doing RLP Encoding of the Header(MixHash, Nonce)
2. We are changing the header.ExtraData after we have computed block hash. When the proposer is building the block, he is in charge of computing the block hash. But when we are inserting the block, we are adding extra info in the header.ExtraData (CommitedSeals). The issue is when go-ethereum client fetches the block, it will have a different header than we had when generating the block hash (because of modified ExtraData). RLP is going to be different so will block hash…

This is tricky to fix, also we are planning to add more info in header.ExtraData (Parant Commited Seals). With different ExtraData our block hash would always be different from the one go-ethereum client gets.

I would like to understand:
1. Where and how are you using this? 
2. Have you found a workaround?

**Comment by scalalang2** [comment]
Thank you for your efforts and invetigation :) 

I wanted to make some communication channel between the offchain server and the blockchain.
If the owner of the NFT on the blockchain is changed to new owner. this changes should be notified to the offchain server.
To make this work, I've been using `ethclient` packages.

At this time, Only `BlockByHash` seems to provides incorrect information.
But, there may be another problem on `ethclient` because other functions were not fully tested

I think all features of the ethclient should be eventually available or **a polygon specific client is required**

**Comment by 0xAleksaOpacic** [comment]
Hey, @scalalang2 
I agree completely. For now, we don't have knowledge that something else is not compatible with ethclient.  

As I said the core issue is that Polygon Edge is calculating block hash differently than the go-ethereum, to be exact we are not using Nonce and MixHash when computing block hashes, as well as the fact that nothing except Validators should be in extraData. You can check the code of how we are computing hash here (https://github.com/0xPolygon/polygon-edge/blob/develop/consensus/ibft/header_hash.go#L11)

Currently, we don't have the capacity to contribute on go-ethereum/ethclient, or to create a fork that modifies how they are calculating block hash to match Polygon Edge.

I am moving this issue to a discussion. In the future, we will consider adding our own client or contributing to go-ethereum/ethclient.  😃
